### PR TITLE
[Refactor] swa-tracker-wrapper

### PR DESCRIPTION
**Pull request by tomer-epstein**

### CODE REVIEW

**Review by slavik-lvovsky — APPROVED** [review]

**Comment by tomer-epstein** [comment]
@ira-gordin-sap appreciate if you can test that it works as expected.

**Review by ira-gordin-sap — CHANGES_REQUESTED** [review]

**Comment by ira-gordin-sap** [comment]
> @ira-gordin-sap appreciate if you can test that it works as expected.

done - works

File: backend/src/swa-tracker/swa-tracker-wrapper.ts [hunk 1]
```diff
@@ -31,78 +31,85 @@ function initSWATracker(newSwaTracker: SWATracker) {
...
- * Note the use of a getter function so the value would be lazy resolved on each use.
- * This enables concise and simple consumption of the swaTracker throughout our Extension.
- *
- * @returns { SWATracker }
- */
-export function getSWA(): SWATracker {
-	if (isInitialized() === false) {
-		throw Error(ERROR_ANALYTICS_TRACKER_NOT_INITIALIZED);
+export class SWA {
+
```
[the remark this answers is not in the document]
Comment (ira-gordin-sap) [reply backend/src/swa-tracker/swa-tracker-wrapper.ts:35]: If you agree, I can push this change.

File: backend/src/swa-tracker/swa-tracker-wrapper.ts [hunk 2]
```diff
@@ -20,89 +20,97 @@ const EVENT_TYPES = {
 
 const ERROR_ANALYTICS_TRACKER_NOT_INITIALIZED = 'Analytics tracker has not yet been initialized!';
```
Comment (ira-gordin-sap) [inline backend/src/swa-tracker/swa-tracker-wrapper.ts:21]: We can put all the constants inside the class as well.

File: backend/src/swa-tracker/swa-tracker-wrapper.ts [hunk 3]
```diff
@@ -20,89 +20,97 @@ const EVENT_TYPES = {
...
+	}
+	
+	
+	public static updateGeneratorStarted(generatorName: string) {
+		const eventType = EVENT_TYPES.PROJECT_GENERATION_STARTED;
+		let customEvents = [generatorName];
+		let startTime = this.startTime = Date.now();
+		this.getSWA().track(eventType, customEvents);
+		getLogger().trace("SAP Web Analytics tracker was called and start time was initialized", {
+			eventType, generatorName, startTime, customEvents});
```
Comment (ira-gordin-sap) [inline backend/src/swa-tracker/swa-tracker-wrapper.ts:79]: We don't need local startTime variable.
Inside the json sent to logger: startTime: this.startTime

File: backend/src/swa-tracker/swa-tracker-wrapper.ts [hunk 4]
```diff
@@ -20,89 +20,97 @@ const EVENT_TYPES = {
...
+		const startTime = this.startTime;
+		const generationTimeMilliSec = endTime - startTime;
+		const generationTimeSec = Math.round(generationTimeMilliSec/1000);
+		let customEvents = [generatorName, generationTimeSec.toString()];
+		if (!_.isNil(errorMessage)) {
+			customEvents.push(errorMessage);
+		}
+		this.getSWA().track(eventType, customEvents);
+		getLogger().trace("SAP Web Analytics tracker was called", 
+			{eventType, generatorName, generationTimeSec, generationTimeMilliSec, endTime, startTime, customEvents, errorMessage});
```
Comment (ira-gordin-sap) [inline backend/src/swa-tracker/swa-tracker-wrapper.ts:99]: We don't need local startTime variable.
Inside the json sent to logger: startTime: this.startTime

**Review by ira-gordin-sap — CHANGES_REQUESTED** [review]

**Review by ira-gordin-sap — APPROVED** [review]

**Pull request merged**